File: src/main/java/dev/kensa/sentence/Sentence.java
package dev.kensa.sentence;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Stream;

import static dev.kensa.sentence.Token.Type.Word;

public class Sentence {

    private final List<SentenceToken> tokens;

    public Sentence(List<SentenceToken> tokens) {
        this.tokens = tokens;
    }

    public Stream<SentenceToken> stream() {
        return tokens.stream();
    }

    public Stream<SentenceToken> squashedTokens() {
        List<SentenceToken> squashed = new ArrayList<>();
        Token.Type currentTokenType = null;
        String currentValue = "";

        for (SentenceToken token : tokens) {
            if (token.type() == Word) {
                if (currentTokenType == Word) {
                    currentValue += " " + token.value();
                } else {
                    currentValue += token.value();
                }
            } else {
                if (currentTokenType == Word) {
                    squashed.add(new SentenceToken(currentTokenType, currentValue));
                    currentValue = "";
                }
                squashed.add(token);
            }
            currentTokenType = token.type();
        }

        if (currentTokenType == Word) {
            squashed.add(new SentenceToken(currentTokenType, currentValue));
        }

        return squashed.stream();
    }
}

File: src/main/java/dev/kensa/sentence/Sentence.kt
package dev.kensa.sentence

import dev.kensa.parse.EmphasisDescriptor
import dev.kensa.parse.HighlightDescriptor
import dev.kensa.sentence.TokenType.Word
import java.util.*

class Sentence(val tokens: List<SentenceToken>) {

    val squashedTokens: List<SentenceToken>
        get() = ArrayList<SentenceToken>().apply {
            var currentTokenTypes: Set<TokenType> = emptySet()
            var currentValue = ""
            var currentEmphasis = EmphasisDescriptor.Default
            var currentHighlight: HighlightDescriptor? = null

            fun finishCurrentWord() {
                if (currentTokenTypes.contains(Word)) {
                    add(SentenceToken(currentValue, currentTokenTypes, emphasis = currentEmphasis, highlight = currentHighlight))
                    currentValue = ""
                    currentEmphasis = EmphasisDescriptor.Default
                    currentHighlight = null
                }
            }

            tokens.forEach { token ->
                if (token.hasType(Word)) {
                    currentValue = if (currentTokenTypes.contains(Word)) {
                        if (currentEmphasis == token.emphasis && currentHighlight == token.highlight) {
                            "$currentValue ${token.value}"
                        } else {
                            finishCurrentWord()
                            token.value
                        }
                    } else {
                        token.value
                    }
                } else {
                    finishCurrentWord()
                    add(token)
                }
                currentTokenTypes = token.tokenTypes
                currentEmphasis = token.emphasis
                currentHighlight = token.highlight
            }
            finishCurrentWord()
        }
}